lib/initialConfigValidation.js

```diff
@@ -0,0 +1,89 @@
+'use strict'
+
+const Ajv = require('ajv')
+const ajv = new Ajv({ removeAdditional: true, useDefaults: true, coerceTypes: true })
+
+// We add a keyword that allow us to set default values
+ajv.addKeyword('setDefaultValue', {
+  modifying: true,
+  validate: function (schemaParamValue, validatedParamValue, validationSchemaObject, currentDataPath, validatedParamObject, validatedParam) {
+    validatedParamObject[validatedParam] = schemaParamValue
+
+    return true
+  },
+  errors: false
+})
+
+const optsSchema = {
+  type: 'object',
+  additionalProperties: false,
+  properties: {
+    bodyLimit: { type: 'integer' },
+    caseSensitive: { type: 'boolean' },
+    http2: {
+      oneOf: [
+        { type: 'boolean' },
+        { type: 'null' }
+      ]
+    },
+    https: {
+      if: {
+        not: {
+          oneOf: [
+            { type: 'boolean' },
+            { type: 'null' },
+            {
+              type: 'object',
+              additionalProperties: false,
+              required: ['allowHTTP1'],
+              properties: {
+                allowHTTP1: { type: 'boolean' }
+              }
+            }
+          ]
+        }
+      },
+      then: { setDefaultValue: true }
+    },
+    ignoreTrailingSlash: { type: 'boolean' },
+    maxParamLength: { type: 'integer' },
+    onProtoPoisoning: { type: 'string' },
+    pluginTimeout: { type: 'integer' },
+    requestIdHeader: { type: 'string' }
+  }
+}
+
+const validate = ajv.compile(optsSchema)
+
+function validateInitialConfig (options) {
+  // We clone the initial configuration options object
+  let opts = Object.assign({}, options)
```

Comment:
Found another case where this PR modifies the user's options:

```js
const Fastify = require('./')
const fs = require('fs')
const path = require('path')
const options = {
  https: {
    allowHTTP1: true,
    key: fs.readFileSync(path.join(__dirname, '..', 'https', 'fastify.key')),
    cert: fs.readFileSync(path.join(__dirname, '..', 'https', 'fastify.cert'))
  }
}

console.log(options.https) // { allowHTTP1: true, key: <Buffer ... >, cert: <Buffer ... > }

const fastify = Fastify(options)

console.log(options.https) // { allowHTTP1: true }
```

The best way to solve this would be to do a deep clone instead of a shallow clone here.

@mcollina What are your thoughts on doing a deep clone? Is there a good way to do it without [`lodash.clonedeep`](https://www.npmjs.com/package/lodash.clonedeep)?

Reply:
Nice catch again :smile:
I set up a git branch [here](https://github.com/darkgl0w/fastify/tree/initial-config-experimentions) where I address **only** this particular case without lodash and not in the most beautiful way (there is so much room to improvements that I am a bit ashamed to just show it :grimacing: )

As I see it now it raise at least 2 concerns :
1. One side effect of this PR is that it opens the path to the proposal submitted here https://github.com/fastify/fastify/issues/793 and so in order to do things in the most clean way, deep cloning should matter in view of the broad spectrum of init options available to the user.
2. Second to this there is the question of adding a new dependency to fastify (that means depending on something that the team do not maintain itself but that does the job well) or instead should we introduce a new "*tools*" lib (be it as a fastify maintained package or just another js file in fastify lib folder) in wich we would provide functions to eg. deep clone or deep freeze objects.

What do you think about this ? should we open a discussion where all the team would share their thoughts about this ?

Reply:
Let's not reinvent the wheel. Are we sure we don't have a deep clone utility in the production dependency tree already? One usually is _always_ downloaded. 

If we don't, let's go with https://github.com/davidmarkclements/rfdc as it's fast.

Reply:
From what i see, the production dependency tree does not have such a tool, we can trigger something using the `deepmerge` package (that is a dependency required by `fast-json-stringify`) like this `deepmerge({}, options)` but it feels more like a dirty hack and it does not cover all the cases that `rfdc` does.

So if you agree I will go with `rfdc` (with both its `proto` and `circles` options set to `true` maybe ? As it is only triggered one time it should not really be a matter in terms of performance cost).

NB: if you want to double check the production dependency tree here is the output of `npm list --production`:
```sh
fastify@2.0.0
├── abstract-logging@1.0.0
├─┬ ajv@6.9.2
│ ├── fast-deep-equal@2.0.1
│ ├── fast-json-stable-stringify@2.0.0
│ ├── json-schema-traverse@0.4.1
│ └─┬ uri-js@4.2.2
│   └── punycode@2.1.1
├─┬ avvio@6.0.1
│ ├─┬ debug@4.1.1
│ │ └── ms@2.1.1
│ └─┬ fastq@1.6.0
│   └── reusify@1.0.4 deduped
├── bourne@1.1.2
├─┬ fast-json-stringify@1.11.0
│ ├── ajv@6.9.2 deduped
│ └── deepmerge@3.2.0
├─┬ find-my-way@2.0.1
│ ├── fast-decode-uri-component@1.0.0
│ ├─┬ safe-regex2@2.0.0
│ │ └── ret@0.2.2
│ └── semver-store@0.3.0
├── flatstr@1.0.9
├─┬ light-my-request@3.2.0
│ ├── ajv@6.9.2 deduped
│ └── readable-stream@3.1.1 deduped
├─┬ middie@4.0.1
│ ├── path-to-regexp@3.0.0
│ └── reusify@1.0.4
├─┬ pino@5.11.1
│ ├── fast-redact@1.4.4
│ ├── fast-safe-stringify@2.0.6
│ ├── flatstr@1.0.9 deduped
│ ├── pino-std-serializers@2.3.0
│ ├── quick-format-unescaped@3.0.2
│ └─┬ sonic-boom@0.7.3
│   └── flatstr@1.0.9 deduped
├─┬ proxy-addr@2.0.4
│ ├── forwarded@0.1.2
│ └── ipaddr.js@1.8.0
├─┬ readable-stream@3.1.1
│ ├── inherits@2.0.3
│ ├─┬ string_decoder@1.2.0
│ │ └── safe-buffer@5.1.2
│ └── util-deprecate@1.0.2
└── tiny-lru@6.0.1
```